packages/@aws-cdk/aws-appmesh/lib/virtual-router-listener.ts

```diff
@@ -0,0 +1,122 @@
+import * as cdk from '@aws-cdk/core';
+import { CfnVirtualRouter } from './appmesh.generated';
+import { Protocol } from './shared-interfaces';
+
+/**
+ * Represents the properties needed to define Listeners for a VirtualRouter
+ */
+export interface RouterListenerProps {
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port?: number
+}
+
+/**
+ * Properties for a VirtualRouter listener
+ */
+export interface VirtualRouterListenerConfig {
+  /**
+   * Single listener config for a VirtualRouter
+   */
+  readonly listener: CfnVirtualRouter.VirtualRouterListenerProperty;
+}
+
+/**
+ * Represents the properties needed to define listeners for a VirtualRouter
+ */
+export abstract class VirtualRouterListener {
+  /**
+   * Returns an HTTP Listener for a VirtualRouter
+   */
+  public static httpVirtualRouterListener(props: RouterListenerProps = {}): VirtualRouterListener {
+    return new HttpVirtualRouterListener(props);
+  }
+
+  /**
+   * Returns an HTTP2 Listener for a VirtualRouter
+   */
+  public static http2GatewayListener(props: RouterListenerProps = {}): VirtualRouterListener {
+    return new Http2VirtualRouterListener(props);
+  }
+
+  /**
+   * Returns a GRPC Listener for a VirtualRouter
+   */
+  public static grpcGatewayListener(props: RouterListenerProps = {}): VirtualRouterListener {
+    return new GrpcVirtualRouterListener(props);
+  }
+
+  /**
+   * Returns a TCP Listener for a VirtualRouter
+   */
+  public static tcpGatewayListener(props: RouterListenerProps = {}): VirtualRouterListener {
```

Comment:
Is there any chance any of these will have protocol-specific properties later? Because frankly, in its current form, I don't think this abstraction is buying us anything...

Reply:
Virtual Routers specifically do not currently have any protocol specific properties. The main reason I think we should do this change is for consistency. If we have listeners on Virtual Nodes and Virtual Gateways, having listeners for your virtual routers that are defined in a different way seems like poor UX. If you don't think it is worth the change, we can scrap this PR.

Reply:
Consistency is not a bad argument.

How about this. Let's make `port?: number` an optional positional argument to each of the static factory methods in this class, and get rid of `RouterListenerProps` completely. Then, creating a Listener will be as simple as calling `VirtualRouterListener.http(8081)`. I think that's a reasonable compromise between consistency and ease of use.

Thoughts on this idea?

Reply:
Sounds good to me. Simpler for sure. Made the update